src/commands/medusaCodeLens.ts

```diff
@@ -15,11 +22,134 @@ export async function generateMedusaDatasetCommand(documentUri: Uri): Promise<vo
     documentUri: documentUri?.toString(),
   });
 
-  // For now, just show an alert
-  // TODO: Implement actual Medusa dataset generation logic
-  await window.showInformationMessage(
-    `Generate Medusa Dataset clicked for: ${documentUri?.fsPath || "Unknown file"}`,
-  );
+  try {
+    // Read the Avro schema file contents
+    const { content: avroSchemaContent } = await getEditorOrFileContents(documentUri);
+
+    if (!avroSchemaContent.trim()) {
+      await window.showErrorMessage("The Avro schema file is empty.");
+      return;
+    }
+
+    logger.info("Reading Avro schema file", {
+      filePath: documentUri.fsPath,
+      contentLength: avroSchemaContent.length,
+    });
+
+    // Show progress while calling the API
+    const dataset = await window.withProgress(
+      {
+        location: { viewId: "confluent" },
+        title: "Generating Medusa Dataset...",
+        cancellable: false,
+      },
+      () => convertAvroSchemaToDataset(avroSchemaContent),
+    );
+
+    // Save the dataset to a file
+    await saveDatasetToFile(dataset);
+  } catch (error) {
+    logger.error("Failed to generate Medusa dataset", error);
+    const errorMessage = error instanceof Error ? error.message : "Unknown error occurred";
+    await window.showErrorMessage(`Failed to generate Medusa dataset: ${errorMessage}`);
+  }
+}
+
+async function convertAvroSchemaToDataset(avroSchemaContent: string): Promise<DatasetDTO> {
+  // Parse and validate the Avro schema content
+  let parsedSchema;
+  try {
+    parsedSchema = JSON.parse(avroSchemaContent);
+  } catch (parseError) {
+    throw new Error(
+      `Invalid JSON in Avro schema file: ${parseError instanceof Error ? parseError.message : "Unknown parsing error"}`,
+    );
+  }
+
+  const medusaContainer = await getMedusaContainer();
```

Comment:
Future: This section from getting the container through to getting reference to the API client is probably going be repeated a few times, so best to move into a convienence function so as to prevent from being duplicated. Can come later when second caller wants it.